app/src/org/commcare/models/database/user/models/CommCareEntityStorageCache.java

```diff
@@ -231,6 +274,15 @@ public void primeCache(Hashtable<String, AsyncEntity> entitySet, String[][] cach
         }
     }
 
+    /**
+     * Gets all shallow records and it's related graph and delete those records from the cache
+     */
+    public void processShallowRecords() {
```

Comment:
Since this is getting schedule asynchronously, I think this method itself might need to be thread protected or something should prevent simultaneous management of the cache while it's happening.

Reply:
https://github.com/dimagi/commcare-android/pull/2955/commits/4d1e04bf1109939598212b2450870e0982911be5